src/crypto/crypto_hkdf.cc

```diff
@@ -103,20 +103,43 @@ bool HKDFTraits::DeriveBits(
   EVPKeyCtxPointer ctx =
       EVPKeyCtxPointer(EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, nullptr));
   if (!ctx || !EVP_PKEY_derive_init(ctx.get()) ||
-      !EVP_PKEY_CTX_hkdf_mode(ctx.get(),
-                              EVP_PKEY_HKDEF_MODE_EXTRACT_AND_EXPAND) ||
       !EVP_PKEY_CTX_set_hkdf_md(ctx.get(), params.digest) ||
-      !EVP_PKEY_CTX_set1_hkdf_salt(
-          ctx.get(), params.salt.data<unsigned char>(), params.salt.size()) ||
-      !EVP_PKEY_CTX_set1_hkdf_key(
-          ctx.get(),
-          reinterpret_cast<const unsigned char*>(params.key->GetSymmetricKey()),
-          params.key->GetSymmetricKeySize()) ||
       !EVP_PKEY_CTX_add1_hkdf_info(
           ctx.get(), params.info.data<unsigned char>(), params.info.size())) {
     return false;
   }
 
+  if (params.key->GetSymmetricKeySize() != 0) {
+    if (!EVP_PKEY_CTX_hkdf_mode(ctx.get(),
+                                EVP_PKEY_HKDEF_MODE_EXTRACT_AND_EXPAND) ||
+        !EVP_PKEY_CTX_set1_hkdf_salt(
+            ctx.get(), params.salt.data<unsigned char>(), params.salt.size()) ||
+        !EVP_PKEY_CTX_set1_hkdf_key(ctx.get(),
+                                    reinterpret_cast<const unsigned char*>(
+                                        params.key->GetSymmetricKey()),
+                                    params.key->GetSymmetricKeySize())) {
+      return false;
+    }
+  } else {
+    unsigned int len = EVP_MD_size(params.digest);
+    uint8_t tempKey[len];  // NOLINT(runtime/arrays)
+    if (params.salt.size()) {
+      HMAC(params.digest,
+           params.salt.data(),
+           params.salt.size(),
+           nullptr,
+           0,
+           tempKey,
+           &len);
+    } else {
+      HMAC(params.digest, new char[len]{}, len, nullptr, 0, tempKey, &len);
```

Comment:
Leaks memory and should check the return value.

Also.... an all-bits-zero key?

Reply:
> an all-bits-zero key?

That's how the HKDF extract step is defined when salt is not given

Reply:
> Leaks memory

can you expand?

Reply:
`new char[len]{}` <- not freed. You can use a zeroed out `char salt[EVP_MAX_KEY_LENGTH]` instead.

Maybe `EVP_MAX_MD_SIZE` is more appropriate but ultimately they're both 64.

Reply:
If, down the line, a code suggestions appears I'd appreciate it. I'm way out of my comfort zone in these code parts.